Export database of Users

sreejakshetty opened the pull request:
#12

![screenshot from 2016-06-06 16 41 34](https://cloud.githubusercontent.com/assets/14835922/15820223/bdd29aac-2c05-11e6-83a7-cd729f90534c.png)
![screenshot from 2016-06-06 16 41 21](https://cloud.githubusercontent.com/assets/14835922/15820222/bdce4150-2c05-11e6-9332-c8656cd0e2af.png)

sreejakshetty committed:
Added export database feature

sreejakshetty committed:
Added Feature : Export Database of Users

kamishettysreeja25 commented:
#12 tested on online server

hpdang commented:
it doesn't work for me. Can you share the exported file you get? 

![screenshot from 2016-06-07 11 43 49](https://cloud.githubusercontent.com/assets/1658742/15853411/33997b76-2ca5-11e6-9a22-ef78738d2a13.png)
![screenshot from 2016-06-07 11 44 06](https://cloud.githubusercontent.com/assets/1658742/15853410/3398523c-2ca5-11e6-8ac4-bf9241fe7e1c.png)

kamishettysreeja25 commented:
![screenshot from 2016-06-07 15 41 39](https://cloud.githubusercontent.com/assets/14835922/15854333/efdd79d8-2cc6-11e6-8ad3-22bdec4ee5cc.png)
![screenshot from 2016-06-07 15 41 24](https://cloud.githubusercontent.com/assets/14835922/15854334/efde8cc4-2cc6-11e6-8e67-7e598003861f.png)
![screenshot from 2016-06-07 15 40 53](https://cloud.githubusercontent.com/assets/14835922/15854335/f036554e-2cc6-11e6-8ece-8dd365e867c8.png)

kamishettysreeja25 commented:
I am getting the html code with the user data . I will check once why I am getting that error.They are getting inserted into the User table.

hpdang commented:
@DishantK1807 I want to have this feature completed for Milestone 5, please test and finish it up.

DishantK1807 commented:
Will start working on it.

DishantK1807 commented:
@kamishettysreeja25 : You are using functions in your code that are deprecated: 

- mysql_connect()
- mysql_errno()
- mysql_error()
- mysql_fetch_row()
- mysql_field_name()
- mysql_num_fields()
- mysql_query()
- mysql_select_db()

the above-mentioned functions are deprecated and we cannot use them in our code. That's the reason we are getting HTML code along with the file.

DishantK1807 commented:
@kamishettysreeja25 : how did you get the screenshot of the database above?

kamishettysreeja25 commented:
@DishantK1807 Which screenshot are you talking about ?
I am able to extract user data but I am getting html code along with it . screenshot shows the user data.

DishantK1807 commented:
@kamishettysreeja25 : Did you changed the functions?

DishantK1807 commented:
@hpdang : I am not able to find possible error in the code. @hongquan :  can you suggest  me a way out?

hongquan commented:
@kamishettysreeja25 Fix the issue of using deprecated 'mysql_xx` functions first. Reused the things in _includes/mysqli_provider.php_. I cannot test because those `mysql_xx` don't work on my machine.

DishantK1807 committed:
Update admin_import.php

kamishettysreeja25 commented:
@hongquan sure. I will work on this issue . I will try to use the functions in mysqli_provider.php and implement it .@DishantK1807  is also working on the same issue . Should we both work on same issue ?

DishantK1807 commented:
@hongquan : Changed those functions. But after changing them the exported database shows "Unable to load angeltypes" 
![capture](https://cloud.githubusercontent.com/assets/7828295/16141240/183c6008-3477-11e6-867a-8252395ae68a.PNG)

hongquan commented:
I recommend you guys to use PHP's built-in server to discover early the PHP error from the console.

![php -s localhost 5000_101](https://cloud.githubusercontent.com/assets/314607/16141538/efa30f42-3486-11e6-9f83-26036f0baec3.png)


@DishantK1807 : Why do you hard-code the MySQL user and password in `export_xls()`:

```
$Connect = sql_connect("localhost", "root", "","engelsystem")
```
What's about if the website is not setup with those value?

Please read my comment above, where I mention about reusing something in  _includes/mysqli_provider.php_.

Also, don't name variable with capitalize letter. Name it `$mything` instead of `$Mything` or `$myThing`.

hongquan commented:
@kamishettysreeja25  Sister H.Phuc wants you guys to finish this soon, you guys can work together on it. Use private chat to decide who to write code, commit code.

DishantK1807 committed:
removed sql_connect()

removed sql_connect() adn sql_select_db(), as already provided in mysqli_provider.php, engelsystem_provider.php for connecting to the db.

hongquan commented:
Try to fix the error which PHP reported:

```
[Fri Jun 17 13:35:53 2016] PHP Warning:  mysqli::query() expects parameter 1 to be string, array given in /home/quan/Works/FOSSASIA/engelsystem/includes/mysqli_provider.php on line 147
[Fri Jun 17 13:35:53 2016] PHP Notice:  Array to string conversion in /home/quan/Works/FOSSASIA/engelsystem/includes/mysqli_provider.php on line 151
[Fri Jun 17 13:35:53 2016] mysql_provider error: MySQL-query error: Array (0)
#1  sql_error(MySQL-query error: Array (0)
) called at [/home/quan/Works/FOSSASIA/engelsystem/includes/mysqli_provider.php:151]
```

DishantK1807 committed:
error handling

DishantK1807 committed:
error handling

hongquan commented:
@DishantK1807 What is the purpose of the last 2 commits, b1be305, 302cf0e? They don't solve any thing and you still push to repo? Please, at least test on your localhost to make it somehow work, before pushing here. Don't pollute the Git history.

Before pushing, please make your commit neat. Those two actually can combine to one. Learn and  use `git rebase -i` to do. Note:  Do `rebase` only on the commits which you haven't pushed to repo.

Those 2 commit even introduce 1 more error:

```
[Fri Jun 17 14:56:18 2016] PHP Notice:  Undefined variable: sql_connection in /home/quan/Works/FOSSASIA/engelsystem/includes/pages/admin_import.php on line 395
```

Please try to test on your localhost first. Use PHP's builtin server to see error early!

DishantK1807 committed:
error handling

DishantK1807 commented:
Sreeja and I tested the new code on our localservers for errors. The exported file still contains HTML code along with the data.

hongquan commented:
That HTML is from this code in _public/index.php_:

```php
echo template_render('../templates/layout.html', array(
    'theme' => isset($user) ? $user['color'] : $default_theme,
    'title' => $title,
    'atom_link' => ($p == 'news' || $p == 'user_meetings') ? '<link href="' . page_link_to('atom') . (($p == 'user_meetings') ? '&amp;meetings=1' : '') . '&amp;key=' . $user['api_key'] . '" type="application/atom+xml" rel="alternate" title="Atom Feed">' : '',
    'menu' => make_menu(),
    'content' => msg() . $content,
    'header_toolbar' => header_toolbar(),
    'faq_url' => $faq_url,
    'contact_email' => $contact_email,
    'locale' => locale() 
));
```

this code is always run and it inserts HTML content to every HTTP response. You have to add some logic to exclude it from your download page.

And @DishantK1807 , you violate coding style again, why all lines in `export_xls()` out of indentation? And why the file name _Webinfopen.csv_? Is it related any to "export user" feature? Did you just copy and paste from somewhere and too lazy to review what it is doing actually?

DishantK1807 commented:
@hongquan : this code was just to know what the problem was and I forgot about the indentation, sorry for that, and now that I've figured out what the actual problem is, I'll work on finding a solution to this problem and make the code right, according to the system.

DishantK1807 committed:
Export user database

DishantK1807 commented:
![capture](https://cloud.githubusercontent.com/assets/7828295/16171134/87d1b512-3585-11e6-98a9-f23f4fccfcd6.PNG)

User database is now exported without the HTML code.

Tested on the online server.

DishantK1807 commented:
@hongquan: Should I make a new PR? as this PR contains many commits that are unnecessary, or we can merge this PR only?

hongquan commented:
Your last commit should be named "Fix: HTML is inserted in export result". I saw many commits with the same title "export user database".

No need to make new PR. You just need to pay attention to the commits next time.

Please make the file name more specific than "export_data.csv". Which data?

Out of indentation:

![selection_104](https://cloud.githubusercontent.com/assets/314607/16176628/2fcee430-363d-11e6-925a-750aeb233b4a.png)

DishantK1807 committed:
Fix: Indentation and file name

DishantK1807 commented:
@hongquan : Updated.

hongquan commented:
Our coding convention is: "Indentation width is 2 spaces, not tab".

hongquan commented on includes/pages/admin_import.php:
```diff
@@ -81,69 +81,83 @@ function admin_import() {
         }
       }
-      
+      if(isset($_REQUEST['download'])){
```
whitespace

hongquan commented:
@DishantK1807 : Many issues:

- Why you report to browser that the file to be delivered is a XLS file, while the real content is CSV?

- Wrong header data
![selection_106](https://cloud.githubusercontent.com/assets/314607/16183313/9ff4638c-36da-11e6-9fb0-b1441298fe40.png)

- You code causes creating a file _export_users_data.csv_ in _public_ folder. What if many users are using the _export_ feature nearly at the same time? They all write to the same file. How the file will be if many people write to it in the same time?

![selection_107](https://cloud.githubusercontent.com/assets/314607/16183393/6c21ed62-36db-11e6-977f-d803d0f76dc0.png)

DishantK1807 commented:
@hongquan : 
> You code causes creating a file export_users_data.csv in public folder. What if many users are using the export feature nearly at the same time? They all write to the same file. How the file will be if many people write to it at the same time?

I am creating a new file on the server so that the HTML content is not inserted into the exported file, that file is saved in the public directory. Only admin can download the user data, no other user is allowed to do so, many people cannot write to it at the same time. Only admin can. And each time the data is exported, the previous file will be replaced.

> Why you report to browser that the file to be delivered is a XLS file, while the real content is CSV?

I was facing problems while writing directly to the XLS file. So, I write into a CSV file and while downloading I changed the content to XLS. The downloaded file is in XLS format. The file is download to the default downloads folder.
![export](https://cloud.githubusercontent.com/assets/7828295/16183806/b46cc162-36d3-11e6-8a53-b57a84a36bfc.PNG)


> Wrong header data

Will fix this now.


Please check this feature on my online server. 
Link: http://stanfordswi.info/engelsystem/public/

DishantK1807 committed:
Fix: Content header and whitespaces

hongquan commented:
I don't ask why you create the file. I concern that what if many users use the export feature and write to that file the same time. "Many users" here is not "non-admin" users, it is "many people login as "admin" role", for example, when H.Phuc and Mario both login to "admin" from different machines and export almost the same time.

An example of problem is that H.Phuc's process want to write "tom@example.com jerry@example.com" to the file, but then Mario's process comes and write "099 654 321", the content of the file will be "tom@example.com 099 654 321 jerry@example.com"

> I was facing problems while writing directly to the XLS file

You don't need to create XLS file.

> while downloading I changed the content to XLS

It is still CSV content:

![selection_108](https://cloud.githubusercontent.com/assets/314607/16183982/9bff26ba-36e2-11e6-9dcf-91db7d0ec489.png)

hongquan commented:
Btw, with your update, the web still write HTML code to HTTP response. But this line

```php
header("Content-Length: ".filesize($filename));
```

is a trick. It will let browser to cut off the content after the part of CSV content, so that the next part with HTML will be ignored by browser.

DishantK1807 commented:
@hongquan : 
> An example of problem is that H.Phuc's process want to write "tom@example.com jerry@example.com" to the file, but then Mario's process comes and write "099 654 321", the content of the file will be "tom@example.com 099 654 321 jerry@example.com"

got it now. I think in that case, the latest transaction would be saved, there would be some delay in the request between both the servers.
What do you suggest we should do?

DishantK1807 commented:
> You don't need to create XLS file.

Will leave the content as CSV only.

hongquan commented:
Create temporary file (I updated my comment above)

DishantK1807 commented:
@hongquan: There are problems with making temporary files. If I use temporary files, the exported file that is downloaded is empty because the temporary file is not saved file is not saved. 

can we merge this PR as of now. I'll make a commit resolving the other issues, and create a separate issue for it, and work on it later, as we have other issues to complete too.

DishantK1807 committed:
Fix: Use temporary file

DishantK1807 commented:
@hongquan : Updated and tested on online server. Used temporary file as you mentioned.
Kindly review the code.

hongquan commented:
Please fix the indentation https://github.com/fossasia/engelsystem/pull/68#issuecomment-227042567

You are slowing down the progress of this PR because you continue making mistake in coding style, which has been reminded more than once.

hongquan commented on includes/pages/admin_import.php:
```diff
@@ -325,49 +337,83 @@ function prepare_events($file, $shifttype_id, $add_minutes_start, $add_minutes_e
 }
+
+function export_xls() {
+	$filename = $fn = tempnam ('/tmp', '.csv'); //  Temporary File Name
```
Whitespace `tempnam ('/tmp', '.csv')`.
Is the variable `$fn` used anywhere? If not, delete it.

DishantK1807 committed:
Fix: Indentation, whitespaces and variable($fn)

hongquan commented on includes/pages/admin_import.php:
```diff
@@ -382,7 +382,7 @@ function shift_sort($a, $b) {
 
 function export_xls() {
-	$filename = $fn = tempnam ('/tmp', '.csv'); //  Temporary File Name
+	$filename = tempnam('/tmp', '.csv');//  Temporary File Name
```
@DishantK1807 Don't delete the space between the code and the comment. That space is to separate code and comment and make it easier to read the code. In some coding convention, people even require at least 2 spaces here. But Drupal doesn't mention, I will skip it now. But next time, please keep space between code and comment (if those are on the same line).
